tests/api_fastapi/execution_api/routes/test_task_instance.py

```diff
@@ -0,0 +1,232 @@
+        Test that the Task Instance state is updated to running when the Task Instance is in a state where it can be
+        marked as running.
+        """
+
+        ti = create_task_instance(
+            task_id="test_ti_update_state_to_running",
+            state=State.QUEUED,
+            session=session,
+        )
+
+        session.commit()
+
+        response = client.patch(
+            f"/execution/task_instance/{ti.id}/state",
+            json={
+                "state": "running",
+                "hostname": "random-hostname",
+                "unixname": "random-unixname",
+                "pid": 100,
+                "start_date": "2024-10-31T12:00:00Z",
+            },
+        )
+
+        assert response.status_code == 204
+        assert response.text == ""
+
+        # Refresh the Task Instance from the database so that we can check the updated values
+        session.refresh(ti)
+        assert ti.state == State.RUNNING
+        assert ti.hostname == "random-hostname"
+        assert ti.unixname == "random-unixname"
+        assert ti.pid == 100
+        assert ti.start_date.isoformat() == "2024-10-31T12:00:00+00:00"
+
+    def test_ti_update_state_conflict_if_not_queued(self, client, session, create_task_instance):
+        """
+        Test that a 409 error is returned when the Task Instance is not in a state where it can be marked as
+        running. In this case, the Task Instance is first in NONE state so it cannot be marked as running.
+        """
+        ti = create_task_instance(
+            task_id="test_ti_update_state_conflict_if_not_queued",
+            state=State.NONE,
+        )
+        session.commit()
+
+        response = client.patch(
+            f"/execution/task_instance/{ti.id}/state",
+            json={
+                "state": "running",
+                "hostname": "random-hostname",
+                "unixname": "random-unixname",
+                "pid": 100,
+                "start_date": "2024-10-31T12:00:00Z",
+            },
+        )
+
+        assert response.status_code == 409
+        assert response.json() == {
+            "detail": {
+                "message": "TI was not in a state where it could be marked as running",
+                "previous_state": State.NONE,
+                "reason": "invalid_state",
+            }
+        }
+
+        assert session.scalar(select(TaskInstance.state).where(TaskInstance.id == ti.id)) == State.NONE
+
+    @pytest.mark.parametrize(
+        ("state", "end_date", "expected_state", "expected_end_date"),
+        [
+            # For SUCCESS & FAILED states, end_date is required
+            (State.SUCCESS, DEFAULT_END_DATE, State.SUCCESS, DEFAULT_END_DATE),
+            (State.FAILED, DEFAULT_END_DATE, State.FAILED, DEFAULT_END_DATE),
+            # When state is SKIPPED, end_date is optional because
+            # 1) a user can run a task
+            #   and raise AirflowSkipException within the task without specifying the end_date
```

Comment:
The user isn't responsible for sending this request, the execution code in task SDK is, so we could choose to enforce an end date for this case if we want/think it's cleaner.

Reply:
Simplified it and removed this in https://github.com/apache/airflow/pull/43602/commits/fa4f2228a2a057d7252e2d1dda64e3575f4b8b94